File: framework/languages/dotnet/e2e/RegistrationCodec/Server/JsonOnlyPeer/DispatchFilters.cs
using RegistrationCodec.Server.JsonOnlyPeer.Infrastructure;
using Zlink.Framework.Contracts.Handlers;

namespace RegistrationCodec.Server.JsonOnlyPeer;

internal sealed class FirstFilter(EvidenceStore evidence) : IZLinkHandlerFilter
{
    public async ValueTask InvokeAsync(
        ZLinkHandlerInvocation invocation,
        ZLinkHandlerFilterNext next,
        CancellationToken cancellationToken)
    {
        evidence.Add($"filter|name=first|phase=before|packet={invocation.PacketName}");
        await next();
        evidence.Add($"filter|name=first|phase=after|packet={invocation.PacketName}");
    }
}

internal sealed class SecondFilter(EvidenceStore evidence) : IZLinkHandlerFilter
{
    public async ValueTask InvokeAsync(
        ZLinkHandlerInvocation invocation,
        ZLinkHandlerFilterNext next,
        CancellationToken cancellationToken)
    {
        evidence.Add($"filter|name=second|phase=before|packet={invocation.PacketName}");
        await next();
        evidence.Add($"filter|name=second|phase=after|packet={invocation.PacketName}");
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Server/Main/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/main/Handlers/DispatchFilters.kt
package systems.zlink.e2e.kotlin.registrationcodec.main.handlers

import java.util.concurrent.CompletionStage
import systems.zlink.e2e.kotlin.registrationcodec.main.infrastructure.ScenarioState
import systems.zlink.framework.ZLinkHandlerFilter
import systems.zlink.framework.ZLinkHandlerFilterNext
import systems.zlink.framework.ZLinkMessageContext

class FirstOrderFilter(
    private val state: ScenarioState,
) : ZLinkHandlerFilter {
    override fun <T : Any?> invoke(
        context: ZLinkMessageContext,
        next: ZLinkHandlerFilterNext<T>,
    ): CompletionStage<T> {
        record(context, "first-before")
        return next.invoke().whenComplete { _, _ -> record(context, "first-after") }
    }

    private fun record(context: ZLinkMessageContext, step: String) {
        state.record("Filter", context.packetName(), step)
    }
}

class SecondOrderFilter(
    private val state: ScenarioState,
) : ZLinkHandlerFilter {
    override fun <T : Any?> invoke(
        context: ZLinkMessageContext,
        next: ZLinkHandlerFilterNext<T>,
    ): CompletionStage<T> {
        record(context, "second-before")
        return next.invoke().whenComplete { _, _ -> record(context, "second-after") }
    }

    private fun record(context: ZLinkMessageContext, step: String) {
        state.record("Filter", context.packetName(), step)
    }
}
